Final/multinblkserver.cpp

```diff
@@ -0,0 +1,129 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netdb.h>
+#include "ThreadPool.h"
+#include "fdlist.h"
+
+const int PORT_NO = 8080;
+const int BACKLOG_SIZE = 2;
+
+// Prototypes
+void select(int sockfd, fd_set *read_fds_ref, struct node *head);
+bool wasEchoed(fd_set *read_fds_ref, struct node *head);
+void echo(int fd);
+int accept(int sockfd);
+
+void error(const char *msg)
+{
+    perror(msg);
+    exit(0);
+}
+
+int main(int argc, char const *argv[])
+{
+	// Step 1: Set up socket
+    int sockfd = socket(AF_INET, SOCK_STREAM, 0); // Create socket file descriptor for server
+    if (sockfd < 0) 
+        error("ERROR opening socket");
+
+	// Step 2: Bind
+	struct sockaddr_in host;
+	host.sin_family = AF_INET; // IPv4 connection
+	host.sin_port = htons(PORT_NO); // bound to port 8080
+	host.sin_addr.s_addr = INADDR_ANY; // connected to localhost
+	const struct sockaddr *host_ptr = (struct sockaddr*) &host;
+	int bound = bind(sockfd, host_ptr, sizeof(host)); // bind sockfd to port
+	if (bound < 0)
+		error("ERROR binding to socket");
+
+	// Step 3: Listen
+	int listening = listen(sockfd, BACKLOG_SIZE); // The socket will actively listen for incoming connections and queue them up
+	if (listening < 0)
+		error("ERROR listener failed");
+
+	fd_set read_fds;
+	struct node *head = createFDList(0); // Create empty fdlist (file descriptor list)
+	while (1) {
```

Comment:
hmmm, let's actually reduce the levels of indirection here. right now, you have a lot of helper methods that are actually obscuring some of the readability of your code. instead, let's do something like this:

```
LinkedList *open_conn_fds = new LinkedList();
while (true) {
    // Reset read_fds
    fd_set read_fds;
    FD_ZERO(&read_fds);
    FD_SET(sockfd, &read_fds);
    for (node *open_conn_fd = open_conn_fds->begin(); open_conn_fd != open_conn_fds->end(); open_conn_fd = open_conn_fd->next) {
        FD_SET(open_conn_fd->data, &read_fds);
    }

    int maxfd = // Compute max fd

    if (select(maxfd + 1, &read_fds, NULL, NULL, NULL) < 0)
        error("My error message");

    if (FD_SET(sockfd, &read_fds)) {
        accept_connection(sockfd, open_conn_fds);
    } else {
        handle_connection(read_fds, open_conn_fds);
    }
}
```

therefore, you only have 2 helper methods:

```
// Accepts the new connection and adds it to the open connections list
void accept_connection(int sockfd, LinkedList *open_conn_fds);

// Handles the ready connection, and then removes the connection from the open connections list
void handle_connection(fd_set read_fds, LinkedList *open_conn_fds);
```
let's start with refactoring the code to something like this, then we can examine how to properly parallelize it.

Reply:
I kept echo abstracted because it's the thing I wanna change to make the server more useful. Is that ok?